retworkx-core/src/shortest_path/bellman_ford.rs

```diff
@@ -0,0 +1,392 @@
+        negative_cycle,
+        scores,
+        predecessor,
+    } = res;
+
+    // Shortest path is not defined
+    if negative_cycle {
+        return Ok(None);
+    }
+
+    // Build path from predecessors
+    if path.is_some() {
+        for node in graph.node_identifiers() {
+            if scores.get_item(node).is_some() {
+                let mut node_path = Vec::<G::NodeId>::new();
+                let mut current_node = node;
+                node_path.push(current_node);
+                while predecessor[current_node.index()].is_some() {
+                    current_node = predecessor[current_node.index()].unwrap();
+                    node_path.push(current_node);
+                }
+                node_path.reverse();
+                path.as_mut().unwrap().insert(node, node_path);
+            }
+        }
+    }
+
+    Ok(Some(scores))
+}
+
+/// Finds an arbitrary negative cycle in a graph using the Bellman-Ford
+/// algorithm with the SPFA heuristic.
+///
+/// Returns a vector of NodeIds if there are cycles, and [`None`] if there aren't.
+/// The output is an arbitrary cycle with the property that the first node of the cycle is also the first
+/// and last element of the vector.
+/// # Example
+/// ```rust
+/// use retworkx_core::petgraph::Graph;
+/// use retworkx_core::petgraph::prelude::*;
+/// use retworkx_core::dictmap::DictMap;
+/// use retworkx_core::shortest_path::negative_cycle_finder;
+/// use retworkx_core::Result;
+///
+/// let mut graph : Graph<(),i32,Directed>= Graph::new();
+/// let a = graph.add_node(()); // node with no weight
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+/// let d = graph.add_node(());
+/// let e = graph.add_node(());
+/// let f = graph.add_node(());
+/// let g = graph.add_node(());
+/// let h = graph.add_node(());
+/// // z will be in another connected component
+/// let z = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///     (a, b, 1),
+///     (b, c, 1),
+///     (c, d, 1),
+///     (d, a, 1),
+///     (e, f, 1),
+///     (b, e, 1),
+///     (f, g, -4),
+///     (g, h, 1),
+///     (h, e, 1)
+/// ]);
+/// // a ----> b ----> e ----> f
+/// // ^       |       ^       |
+/// // |       v       |       v
+/// // d <---- c       h <---- g
+///
+/// let res: Result<Option<Vec<NodeIndex>>> = negative_cycle_finder(
+///     &graph, |x| Ok(*x.weight())
+/// );
+/// assert_eq!(res.unwrap().unwrap().len() - 1, 4);
+/// // the first/last node in the cycle appears twice
+/// ```
+pub fn negative_cycle_finder<G, F, K, E>(
+    graph: G,
+    edge_cost: F,
+) -> Result<Option<Vec<G::NodeId>>, E>
+where
+    G: IntoEdges + Visitable + NodeIndexable + NodeCount + IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash + IndexType,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+{
+    let starts: Vec<G::NodeId> = graph.node_identifiers().collect();
+    let res: BellmanFordData<G, Vec<Option<K>>> = inner_bellman_ford(graph, starts, edge_cost)?;
```

Comment:
We don't really need to store node ids in a vector. 
```suggestion
    let res: BellmanFordData<G, Vec<Option<K>>> =
        inner_bellman_ford(graph, graph.node_identifiers(), edge_cost)?;
```